airbyte-commons-protocol/src/main/java/io/airbyte/commons/protocol/migrations/MigrationContainer.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.commons.protocol.migrations;
+
+import io.airbyte.commons.version.Version;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.function.BiFunction;
+
+public class MigrationContainer<T extends Migration> {
+
+  private final List<T> migrationsToRegister;
+  private final SortedMap<String, T> migrations = new TreeMap<>();
+  private String mostRecentMajorVersion = "";
+
+  public MigrationContainer(final List<T> migrations) {
+    this.migrationsToRegister = migrations;
+  }
+
+  public void initialize() {
+    migrationsToRegister.forEach(this::registerMigration);
+  }
+
+  public Version getMostRecentVersion() {
+    return new Version(mostRecentMajorVersion, "0", "0");
+  }
+
+  /**
+   * Downgrade a message from the most recent version to the target version by chaining all the
+   * required migrations
+   */
+  public <PreviousVersion, CurrentVersion> PreviousVersion downgrade(final CurrentVersion message,
+                                                                     final Version target,
+                                                                     final BiFunction<T, Object, Object> applyDowngrade) {
+    if (target.getMajorVersion().equals(mostRecentMajorVersion)) {
+      return (PreviousVersion) message;
+    }
+
+    Object result = message;
```

Comment:
Could this be defined as `PreviousVersion results`?

Reply:
No because of the chaining of transforms, if we end up having more than one migration, the type of `results` will change.